src/pam/mod.rs

```diff
@@ -34,103 +35,81 @@ pub struct PamContext<C: Converser> {
-        self.converser = Some(converser);
-        self
-    }
-
-    /// Set the service name for the PAM session.
-    ///
-    /// Note that the service name should be based on a static string and not
-    /// based on the name of the binary.
-    pub fn service_name<T: Into<String>>(mut self, name: T) -> PamContextBuilder<C> {
-        self.service_name = Some(name.into());
-        self
-    }
-
-    /// Set a target user that should be inserted into the pam context.
+impl PamContext<CLIConverser> {
+    /// Build the PamContext with the CLI conversation function.
     ///
     /// The target user is optional and may also be set after the context was
     /// constructed or not set at all in which case PAM will ask for a
     /// username.
-    pub fn target_user<T: Into<String>>(mut self, user: T) -> PamContextBuilder<C> {
-        self.target_user = Some(user.into());
-        self
-    }
-}
+    ///
+    /// This function will error when initialization of the PAM session somehow failed.
+    pub fn new_cli(
+        converser_name: &str,
+        service_name: &str,
+        use_stdin: bool,
+        no_interact: bool,
+        password_feedback: bool,
+        target_user: Option<&str>,
+    ) -> PamResult<PamContext<CLIConverser>> {
+        let converser = CLIConverser {
+            name: converser_name.to_owned(),
+            use_stdin,
+            no_interact,
+            password_feedback,
+        };
 
-impl<C> Default for PamContextBuilder<C> {
-    fn default() -> Self {
-        Self {
-            converser: None,
-            service_name: None,
-            target_user: None,
-        }
+        Self::new(converser, service_name, target_user)
     }
 }
 
 impl<C: Converser> PamContext<C> {
+    fn new(
+        converser: C,
+        service_name: &str,
+        target_user: Option<&str>,
+    ) -> PamResult<PamContext<C>> {
+        let c_service_name = CString::new(service_name)?;
+        let c_user = target_user.map(CString::new).transpose()?;
+        let c_user_ptr = match c_user {
+            Some(ref c) => c.as_ptr(),
+            None => std::ptr::null(),
+        };
+
+        // this will be de-allocated explicitly in this type's drop method
+        let data_ptr = Box::into_raw(Box::new(ConverserData {
+            converser,
+            panicked: false,
+        }));
+
+        let mut pamh = std::ptr::null_mut();
+        // SAFETY: we are passing the required fields to `pam_start`; in particular, the value
+        // of `pamh` set above is not used, but will be overwritten by `pam_start`.
+        let res = unsafe {
+            pam_start(
+                c_service_name.as_ptr(),
+                c_user_ptr,
+                &pam_conv {
+                    conv: Some(converse::converse::<C>),
+                    appdata_ptr: data_ptr as *mut libc::c_void,
+                },
+                &mut pamh,
+            )
+        };
+
+        pam_err(res)?;
+
+        assert!(!pamh.is_null());
```

Comment:
Same question as before: I'm guessing the reason for this assert is that PAM guarantees we have a non-`NULL` pamh at this point? If so, comment can be resolved :-)

Reply:
The docs don't explicitly mention this, but og sudo doesn't check for NULL at all, so if PAM ever doesn't set pamh without returning an error, that would crash og sudo too.

Reply:
That would be an assumption that I would make in sudo's place as well. Fair enough.